air/src/proof/mod.rs

```diff
@@ -207,61 +217,276 @@ fn get_conjectured_security(
+    let fri_commit_err_bits = extension_field_bits
+        - ((0.5 * (m + 0.5).powf(7.0) / rho.powf(1.5)) * lde_domain_size.powf(2.0)).log2();
+
+    // Compute FRI query-phase soundness error
+    let fri_queries_err_bits =
+        options.grinding_factor() as f64 - ((1.0 - theta).powf(num_fri_queries)).log2();
+
+    // Combined error for FRI
+    let fri_err_bits = cmp::min(fri_commit_err_bits as u64, fri_queries_err_bits as u64);
+    if fri_err_bits < 1 {
+        return 0;
+    }
+    let fri_err_bits = fri_err_bits - 1;
+
+    // To apply Theorem 8 in https://eprint.iacr.org/2022/1216.pdf, we need to apply FRI with
+    // a slightly larger agreement parameter alpha.
+    // More concretely, we need alpha > rho_plus.sqrt() where rho_plus is the rate in function field
+    // F(Z) and defined as (trace_domain_size + 2.0) / lde_domain_size .
+    // This means that the range of m needs to be restricted in order to insure that
+    // alpha := 1 - theta := rho.sqrt() * (1 + 1/2m) is greater than rho_plus.sqrt().
+    // Determining the range of m is the responsibility of the calling function.
+    // Now, once m is fixed, we need to make sure that we choose an m_plus such that
+    // alpha <= rho_plus.sqrt() * (1 + 1/2m_plus). This m_plus will be used to define
+    // the list-decoding list size in F(Z).
+
+    // Modified rate in function field F(Z)
+    let rho_plus = (trace_domain_size + 2.0) / lde_domain_size;
+    // New proximity parameter m_plus, corresponding to rho_plus, needed to make sure that
+    //  alpha < rho_plus.sqrt() * (1 + 1 / (2 * m_plus))
+    let m_plus = (1.0 / (2.0 * (alpha / rho_plus.sqrt() - 1.0))).ceil();
+
+    // List size
+    let l_plus = (2.0 * m_plus + 1.0) / (2.0 * rho_plus.sqrt());
+
+    // ALI related soundness error. Note that C here is equal to 1 because of our use of
+    // linear batching.
+    let ali_err_bits = -l_plus.log2() + extension_field_bits;
+
+    // DEEP related soundness error. Note that this uses that the denominator |F| - |D ∪ H|
+    // can be approximated by |F| for all practical domain sizes. We also use the blow-up factor
+    // as an upper bound for the maximal constraint degree.
+    let deep_err_bits =
+        -(l_plus * (max_deg * (trace_domain_size + 1.0) + (trace_domain_size - 1.0))).log2()
+            + extension_field_bits;
+
+    let min = cmp::min(
+        cmp::min(fri_err_bits, ali_err_bits as u64),
+        deep_err_bits as u64,
+    );
+    if min < 1 {
+        return 0;
+    }
 
-    // compute security we get by executing multiple query rounds
-    let security_per_query = 0.5 * blowup_plus_bits - (1.0 + 1.0 / (2.0 * m)).log2();
-    let mut query_security = (security_per_query * num_fri_queries) as u32;
+    min - 1
+}
+
+/// Computes proven security level for the specified proof parameters for a fixed
+/// value of the proximity parameter m.
+#[cfg(not(feature = "std"))]
+fn proven_security_protocol_m_nostd(
```

Comment:
Question: instead of creating a separate version of this entire function, could we create separate version of just `log2()` function? For example:

```Rust
#[cfg(feature = "std")]
fn log2(value: f64) -> f64 {
    value.log2()
}

#[cfg(not(feature = "std"))]
fn log2(value: f64) -> f64 {
    // our custom implementation
}
```

Reply:
Makes total sense! I went ahead and did the same for the rest of the operations we make use of. Let me know if the approach agrees with what you proposed.